cli/internal/oauth/manager/manager.go

```diff
@@ -0,0 +1,217 @@
+package manager
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/docker/cli/cli/config/credentials"
+	"github.com/docker/cli/cli/config/types"
+	"github.com/docker/cli/cli/internal/oauth/api"
+	"github.com/docker/cli/cli/internal/oauth/util"
+	"github.com/sirupsen/logrus"
+)
+
+// OAuthManager is the manager
+type OAuthManager struct {
+	api         api.OAuthAPI
+	audience    string
+	tenant      string
+	credStore   credentials.Store
+	openBrowser func(string) error
+}
+
+// OAuthManagerOptions is the options used for New to create a new auth manager.
+type OAuthManagerOptions struct {
+	Audience    string
+	ClientID    string
+	Scopes      []string
+	Tenant      string
+	DeviceName  string
+	Store       credentials.Store
+	OpenBrowser func(string) error
+}
+
+// TokenResult is a result from the auth manager.
+type TokenResult struct {
+	AccessToken  string
+	RefreshToken string
+	RequireAuth  bool
+	Tenant       string
+	Claims       util.Claims
+}
+
+func New(options OAuthManagerOptions) (*OAuthManager, error) {
+	scopes := []string{"openid", "offline_access"}
+	if len(options.Scopes) > 0 {
+		scopes = options.Scopes
+	}
+
+	openBrowser := util.OpenBrowser
+	if options.OpenBrowser != nil {
+		openBrowser = options.OpenBrowser
+	}
+
+	return &OAuthManager{
+		audience: options.Audience,
+		api: api.API{
+			BaseURL:  "https://" + options.Tenant,
+			ClientID: options.ClientID,
+			Scopes:   scopes,
+			Client: util.Client{
+				UserAgent: options.DeviceName,
+			},
+		},
+		tenant:      options.Tenant,
+		credStore:   options.Store,
+		openBrowser: openBrowser,
+	}, nil
+}
+
+// LoginDevice launches the device authentication flow with the tenant, printing instructions
+// to the provided writer and attempting to open the browser for the user to authenticate.
+// Once complete, the retrieved tokens are stored and returned.
+func (m *OAuthManager) LoginDevice(w io.Writer) (res TokenResult, err error) {
+	state, err := m.api.GetDeviceCode(m.audience)
+	if err != nil {
+		return res, fmt.Errorf("failed to get device code: %w", err)
+	}
+
+	if state.UserCode == "" {
+		return res, errors.New("login failed: no user code returned")
+	}
+
+	// todo(laurazard): review the copy here
+	_, _ = fmt.Fprintf(w, "Attempting to automatically open the login authorization page in your default browser.\nIf the browser does not open or you wish to use a different device to authorize this request, open the following URL:\n\n%s\n", strings.Split(state.VerificationURI, "?")[0])
+	_ = m.openBrowser(state.VerificationURI)
+
+	_, _ = fmt.Fprintf(w, "\nAnd enter the device confirmation code: %s\n", state.UserCode)
+	_, _ = fmt.Fprint(w, "\nWaiting for authentication in the browser...\n")
```

Comment:
Maybe something like this... if we have a "press enter to confirm" before launching the browser.

```suggestion
	_, _ = fmt.Fprintln(w, "You will be signed in using a web-based login.")
	_, _ = fmt.Fprintln(w, "To sign in with credentials on the command line, use 'docker login -u <username>'")
	_, _ = fmt.Fprintf(w, "\nYour one-time device confirmation code is: %s\n", state.UserCode)
	reader := bufio.NewReader(os.Stdin)
	_, _ = fmt.Fprint(w, "\nPress ENTER to open the browser.\n")
	_, _ = fmt.Fprintf(w, "Or open the URL manually: %s.\n", strings.Split(state.VerificationURI, "?")[0])
	reader.ReadString('\n')
	_, _ = fmt.Fprint(w, "\nWaiting for authentication in the browser...\n")
```

Reply:
beautiful!